File: app/src/main/java/com/iita/akilimo/utils/FertilizerList.java
package com.iita.akilimo.utils;

import android.os.Build;

import com.iita.akilimo.models.Fertilizer;

import java.util.Iterator;
import java.util.List;

public class FertilizerList {


    @Deprecated
    public static List<Fertilizer> filterSelectedFertilizers(List<Fertilizer> fertilizerTypeList, List<Fertilizer> selectedFertilizers) {
        for (Fertilizer selectedFertilizer : selectedFertilizers) {
            Iterator itr = fertilizerTypeList.iterator();
            while (itr.hasNext()) {
                Fertilizer x = (Fertilizer) itr.next();
                if (x.getName().equalsIgnoreCase(selectedFertilizer.getName())) {
                    itr.remove(); //remove from fertilizer name list to prevent duplicate when we merge them
                }
            }
        }
        fertilizerTypeList.addAll(selectedFertilizers);

        return fertilizerTypeList;
    }

    public static List<Fertilizer> removeFertilizerByType(List<Fertilizer> fertilizerTypeList, final String fertilizerTypeToRemove) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            fertilizerTypeList.removeIf(obj -> obj.getType().equalsIgnoreCase(fertilizerTypeToRemove));
        } else {
            //legacy android version
            for (Iterator<Fertilizer> iterator = fertilizerTypeList.iterator(); iterator.hasNext(); ) {
                if (iterator.next().getType().equalsIgnoreCase(fertilizerTypeToRemove)) {
                    iterator.remove();
                }
            }
        }

        return fertilizerTypeList;
    }
}

File: app/src/main/java/com/iita/akilimo/utils/FertilizerList.kt
package com.iita.akilimo.utils

import android.os.Build

import com.iita.akilimo.entities.Fertilizer
import com.iita.akilimo.entities.InterCropFertilizer
import java.util.stream.Collectors




object FertilizerList {

    fun removeFertilizerByType(
        fertilizerTypeList: MutableList<Fertilizer>,
        fertilizerTypeToRemove: String
    ): List<Fertilizer> {
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.N -> fertilizerTypeList.removeIf { obj ->
                obj.fertilizerType.equals(fertilizerTypeToRemove, ignoreCase = true)
            }
            else -> {
                val iterator = fertilizerTypeList.iterator()
                while (iterator.hasNext()) {
                    if (iterator.next().fertilizerType.equals(
                            fertilizerTypeToRemove,
                            ignoreCase = true
                        )
                    ) {
                        iterator.remove()
                    }
                }
            }
        }

        return fertilizerTypeList
    }


    fun removeIntercropFertilizerByType(
        fertilizerTypeList: MutableList<InterCropFertilizer>,
        fertilizerTypeToRemove: String
    ): List<InterCropFertilizer> {
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.N -> fertilizerTypeList.removeIf { obj ->
                obj.fertilizerType.equals(fertilizerTypeToRemove, ignoreCase = true)
            }
            else -> {
                val iterator = fertilizerTypeList.iterator()
                while (iterator.hasNext()) {
                    if (iterator.next().fertilizerType.equals(
                            fertilizerTypeToRemove,
                            ignoreCase = true
                        )
                    ) {
                        iterator.remove()
                    }
                }
            }
        }

        return fertilizerTypeList
    }
}
